### PR TITLE
feat: fall-detect initialization

### PR DESCRIPTION
**Pull request by bhavikapanara**
This PR is initialization of stand-alone fall detect python package

Run python file for fall-detection: `python3 demo-fall-detection.py `

To test fall-detection using the command line for 2 images:
`python3 demo-fall-detection-cmd.py --image_1 Images/fall_img_1.png --image_2 Images/fall_img_2.png
` 

To test fall-detection using the command line for 3 images:
`python3 demo-fall-detection-cmd.py --image_1 Images/fall_img_1.png --image_2 Images/fall_img_2.png --image_3 Images/fall_img_3.png 
` 

Use `Demo.ipynb` jupyter-notebook for the experiment.

### CODE REVIEW

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara you've copied virtually all the code from the ambianic-edge repo. This is not at all the goal here.

Please remove all files and refactor fall detection classes as needed. If there is any code that is not directly involved in the fall detection algorithm, then it does not belong in this repo.

You should also have a very clear plan and design how ambianic-edge will import and use the fall-detection python package.

Please prepare a design document so we can align on high level goals. I usually use diagrams.net for design diagrams but you can use any similar tool that you like.

**Comment by bhavikapanara** [comment]
> @bhavikapanara you've copied virtually all the code from the ambianic-edge repo. This is not at all the goal here.
> 
> Please remove all files and refactor fall detection classes as needed. If there is any code that is not directly involved in the fall detection algorithm, then it does not belong in this repo.
> 
> You should also have a very clear plan and design how ambianic-edge will import and use the fall-detection python package.
> 
> Please prepare a design document so we can align on high level goals. I usually use diagrams.net for design diagrams but you can use any similar tool that you like.

@ivelin Yes...I have tried to consider the only code that is directly involved in fall detection. However, again I will recheck it and refactor fall detection classes if require.

And, Will prepare the design document and get back to you.

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara this looks better.

There is only one file that seems unnecessary: .peerjsrc

Can you please share your forked repo for this PR with CI passing so I can see what the master repo would look like after merge?

**Comment by ivelin** [comment]
@bhavikapanara I don't see a README.md for the repo. Why was it deleted?

File: LICENSE [hunk 1]
```diff
@@ -1,201 +0,0 @@
-                                 Apache License
```
Comment (ivelin) [inline LICENSE:1]: @bhavikapanara Why remove the LICENSE file?

File: README.md [hunk 2]
```diff
@@ -1,52 +0,0 @@
-# People Fall Detection
```
Comment (ivelin) [inline README.md:1]: @bhavikapanara why delete the README.md. It needs to be filled out with more relevant content.

File: README.md [hunk 3]
```diff
@@ -1,52 +0,0 @@
...
-
-_This repo is the new home of the fall detection model used in Ambianic Edge. We are preparing to move the code over from the [ambianic-edge repo](https://github.com/ambianic/ambianic-edge/blob/master/src/ambianic/pipeline/ai/fall_detect.py) and make it available as a standalone library._
-
-# TODO
-
```
Comment (ivelin) [inline README.md:6]: @bhavikapanara check off the boxes in this list that are done with this PR.

File: README.md [hunk 4]
```diff
@@ -1,52 +0,0 @@
...
-- [ ] Standalone Python ML library for people fall detection based on Tensorflow and [PoseNet 2.0](https://github.com/tensorflow/tfjs-models/tree/master/posenet).
-- [ ] Python source code and wheel package published on PyPi 
-- [ ] Jupyter Notebook to interactively test and experiement with the model
-- [ ] CI & test suite
-- [ ] Training and testing data sets
-- [ ] Third party ML models used as building blocks
```
Comment (ivelin) [inline README.md:12]: @bhavikapanara this README needs to have sections on basic usage for library API, jupyter notebook usage and CLI usage.

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara commented on a few more areas that need work.

File: .peerjsrc [hunk 5]
```diff
@@ -0,0 +1 @@
+{"peerId": "497c5e67-c347-4663-baec-fc7a0fa29a41"}
```
[the remark this answers is not in the document]
Comment (bhavikapanara) [reply .peerjsrc:1]: yes...it's not required here.

**Comment by bhavikapanara** [comment]
> @bhavikapanara I don't see a README.md for the repo. Why was it deleted?

Ohh...by mistake it deleted. you can see it in the next update.

File: README.md (hunk 4 above)
Comment (bhavikapanara) [reply README.md:12]: added

File: LICENSE (hunk 1 above)
Comment (bhavikapanara) [reply LICENSE]: By mistake, it was removed. you can check in the current update.

**Comment by bhavikapanara** [comment]
> @bhavikapanara commented on a few more areas that need work.

tried to cover all comments in the latest update...Please review it and let me know if I missed anything.

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara nice progress.

Next, lets work on the following items:
- port fall detector tests from edge to this repo
- write a github action with the following jobs:
  - build
  - run tests
  - run semantic release to push a pypi package as needed

You can borrow most of the steps from the [peerjs-python github action](https://github.com/ambianic/peerjs-python/tree/master/.github/workflows).

The semantic release section is now available via [more compact github action plugin](https://github.com/ambianic/ambianic-edge/blob/440cebc2b53da25d7cbb07c873bc1cf603010640/.github/workflows/main.yml#L103) that we use in the edge CI.

File: .github/workflows/pythonpackage.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    runs-on: ubuntu-20.04
+    strategy:
+      max-parallel: 4
+      matrix:
+        python-version: [3.7, 3.8]
+
+    steps:
+    - uses: actions/checkout@v1
+    - name: Install aiortc native dependencies
```
Comment (ivelin) [inline .github/workflows/pythonpackage.yml:17]: @bhavikapanara we don't need aiortc for this package. That's a webrtc p2p communication library.

File: README.md [hunk 7]
```diff
@@ -4,9 +4,9 @@ _This repo is the new home of the fall detection model used in Ambianic Edge. We
 
 # TODO
 
-- [ ] Standalone Python ML library for people fall detection based on Tensorflow and [PoseNet 2.0](https://github.com/tensorflow/tfjs-models/tree/master/posenet).
+- [x] Standalone Python ML library for people fall detection based on Tensorflow and [PoseNet 2.0](https://github.com/tensorflow/tfjs-models/tree/master/posenet).
 - [ ] Python source code and wheel package published on PyPi 
-- [ ] Jupyter Notebook to interactively test and experiement with the model
+- [x] Jupyter Notebook to interactively test and experiement with the model
 - [ ] CI & test suite
```
Comment (ivelin) [inline README.md:10]: @bhavikapanara any reason the CI & test suite checkbox is not checked yet?

File: README.md [hunk 8]
```diff
@@ -4,9 +4,9 @@ _This repo is the new home of the fall detection model used in Ambianic Edge. We
...
 - [ ] Training and testing data sets
```
Comment (ivelin) [inline README.md:11]: @bhavikapanara why isn't the test data check box checked?

File: README.md [hunk 9]
```diff
@@ -4,9 +4,9 @@ _This repo is the new home of the fall detection model used in Ambianic Edge. We
...
 - [ ] Third party ML models used as building blocks
```
Comment (ivelin) [inline README.md:12]: @bhavikapanara the posenet tflite models are included in this PR , correct? This box needs to be checked.

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+PyYAML>=5.1.2
+requests>=2.25
+requests-oauthlib>=1.0.0
+responses>=0.9.0
+simplegeneric>=0.8.1
+simplejson>=3.16.0
+webencodings>=0.5.1
+Werkzeug>=0.15.3
+concurrent-log-handler>=0.9.19
+watchdog>=0.10
```
Comment (ivelin) [inline requirements.txt:38]: @bhavikapanara please clean up this requirements.txt list. May libs are not required for the standalone fall detection library.

File: src/__init__.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
+import os
+from dynaconf import Dynaconf
```
Comment (ivelin) [inline src/__init__.py:2]: @bhavikapanara we don't need all these configuration files for a standalone library.

File: src/pipeline/__init__.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
+"""Main module for Ambianic AI pipelines."""
```
Comment (ivelin) [inline src/pipeline/__init__.py:1]: @bhavikapanara pipelines should not be part of a standalone fall detection library.

File: src/util.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
+"""Service classes for OS interaction and multithreading."""
```
Comment (ivelin) [inline src/util.py:1]: @bhavikapanara most of the code in this util.py file is not required for this repo. Please clean up unused code.

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara good progress. Some cleanup still required. See inline comments.

File: src/util.py (hunk 13 above)
Comment (bhavikapanara) [reply src/util.py:1]: done

File: src/pipeline/__init__.py (hunk 12 above)
Comment (bhavikapanara) [reply src/pipeline/__init__.py:1]: done

File: src/__init__.py (hunk 11 above)
Comment (bhavikapanara) [reply src/__init__.py:2]: done

File: README.md (hunk 9 above)
Comment (bhavikapanara) [reply README.md:12]: done

File: README.md (hunk 8 above)
Comment (bhavikapanara) [reply README.md:11]: done

File: README.md (hunk 7 above)
Comment (bhavikapanara) [reply README.md:10]: done

File: .github/workflows/pythonpackage.yml (hunk 6 above)
Comment (bhavikapanara) [reply .github/workflows/pythonpackage.yml:17]: done

File: requirements.txt (hunk 10 above)
Comment (bhavikapanara) [reply requirements.txt]: done...
In case if I missed to remove any unnecessary packages from the requirements.txt list please let me know.

**Review by ivelin — CHANGES_REQUESTED** [review]
Some work remains on dependency cleanup and common code reuse between repos.

File: src/pipeline/tf_detect.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
+"""Tensorflow image detection wrapper."""
```
[the remark this answers is not in the document]
Comment (bhavikapanara) [reply src/pipeline/tf_detect.py]: @ivelin I have updated the code...please take a look at it.

File: src/setup.cfg [hunk 15]
```diff
@@ -0,0 +1,24 @@
+[metadata]
+name = ambianic-fall-detection
+version = 1.0.0
+author = Ivelin Ivanov
```
Comment (ivelin) [inline src/setup.cfg:4]: @bhavikapanara add your name to the authors list

File: requirements.txt [hunk 16]
```diff
@@ -0,0 +1,13 @@
+absl-py>=0.7.1
+appdirs>=1.4.3
+gast>=0.2.2
+h5py>=2.9.0
+matplotlib==3.3.2
+numpy>=1.16.2
+oauthlib>=2.1.0
+Pillow>=5.4.1
+PyYAML>=5.1.2
+requests>=2.25
```
Comment (ivelin) [inline requirements.txt:10]: @bhavikapanara do we need `requests, requests-oauthlib, responses, simplejson, oauthlib, matplotlib, h5py, gast, appdirs, absl-py` ?

Please do a thorough cleanup to ensure there are no unused packages in requirements.txt.

**Review by ivelin — CHANGES_REQUESTED** [review]
See additional comments.

Please comment how you think about using this standalone library in ambianic-edge. What is the API that will be used by edge for this model and potentially other standalone AI models?

File: src/setup.cfg (hunk 15 above)
Comment (bhavikapanara) [reply src/setup.cfg]: done

File: requirements.txt (hunk 16 above)
Comment (bhavikapanara) [reply requirements.txt]: done

**Comment by bhavikapanara** [comment]
> See additional comments.
> 
> Please comment how you think about using this standalone library in ambianic-edge. What is the API that will be used by edge for this model and potentially other standalone AI models?


we can use the `Fall_prediction` method `(exist in fall_prediction.py)` with pipeline image 2 or 3. Similar to the demo example `demo-fall-detection.py`. I think this is one way to call this fall-detect stand-alone package in ambianic-edge.

Please guide me if any other good method that we can implement.

**Comment by ivelin** [comment]
OK, can you please show me a few lines of pseudo code. Which class in the
ambianic edge will interact with the fall detector package?

**Comment by bhavikapanara** [comment]
after brainstorming with the goal I found one possible way to integrate an ambianic-edge with the fall-detect module.

We can directly attach an ambianic-edge pipeline with a fall-detect module and call the fall_detection process on the `receive_next_sample()` method of `PipeElement`


`    def receive_next_sample(self, **sample):`
[https://github.com/ambianic/ambianic-edge/blob/0291b2abfd76fd9a985a5b09445a505dfd4f83ca/src/ambianic/pipeline/__init__.py#L163](url)

Here is an example to call fall-detect module:
[https://github.com/ambianic/fall-detection/blob/06894c0d3660049b027eaa57a8167ce68824de8b/fall_prediction.py#L43](url)

**Comment by ivelin** [comment]
> after brainstorming with the goal I found one possible way to integrate an ambianic-edge with the fall-detect module.
> 
> We can directly attach an ambianic-edge pipeline with a fall-detect module and call the fall_detection process on the `receive_next_sample()` method of `PipeElement`
> 
> [` def receive_next_sample(self, **sample):`](https://github.com/ambianic/ambianic-edge/blob/0291b2abfd76fd9a985a5b09445a505dfd4f83ca/src/ambianic/pipeline/__init__.py#L163)
> 
> Here is an example to call fall-detect module: [Link](https://github.com/ambianic/fall-detection/blob/06894c0d3660049b027eaa57a8167ce68824de8b/fall_prediction.py#L43)

Yes, that seems like a sensible approach.

**Review by ivelin — CHANGES_REQUESTED** [review]
@bhavikapanara I don't see codecoverage reporting in the CI.

**Review by ivelin — APPROVED** [review]
@bhavikapanara 
OK, let's give it a shot. Please be on standby for potential issues we have not caught in the review process.

**Comment by bhavikapanara** [comment]
> @bhavikapanara
> OK, let's give it a shot. Please be on standby for potential issues we have not caught in the review process.

yes...sure @ivelin 
Thanks

**Pull request merged**